php/libraries/VisitTest.php

```diff
@@ -0,0 +1,380 @@
+            ) == false
+        ) {
+            $error = true;
+        }
+
+        $param = array(
+                  'projectID'    => 2,
+                  'subprojectID' => 12,
+                 );
+        Visit::deleteFromProject($this->idList[3], $param);
+        $result = Visit::getList($param);
+        if ($this->_assertArray($result, array()) == false) {
+            $error = true;
+        }
+
+        $param = array('subprojectID' => 11);
+        Visit::deleteFromProject($this->idList[2], $param);
+        $result = Visit::getList($param);
+        if ($this->_assertArray($result, array()) == false) {
+            $error = true;
+        }
+
+        $param  = array();
+        $result = Visit::deleteFromProject($this->idList[0], $param);
+        if ($result != false) {
+            $error = true;
+        }
+
+        return $error;
+    }
+
+    /**
+     * Check if array are equals
+     *
+     * @param array $array1 the first array
+     * @param array $array2 the second array
+     *
+     * @return boolean true if there are equals
+     */
+    private function _assertArray($array1, $array2)
+    {
+        // check for phpunit function call
+        if (sizeof($array1) != sizeof($array2)) {
+            print("Array not the same size\n");
+            return false;
+        }
+
+        foreach ($array1 as $key=>$value) {
+            if (isset($array2[$key]) && is_array($array2[$key])) {
+                $result = $this->_assertArray($value, $array2[$key]);
+                if ($result == false) {
+                    return false;
+                }
+            } elseif (!(isset($array2[$key]) && $value === $array2[$key]
+                || is_null($value) && is_null($array2[$key]) )
+            ) {
+                print("Value are not the same\n");
+                return false;
+            }
+        }
+        return true;
+    }
+}
+
+// prepare Database for test;
+global $DB ;
+$DB = new Database();
+$DB = Database::singleton($database, $username, $password, $host);
+$DB->run("DELETE FROM visit_project_subproject_rel");
+$DB->run("DELETE FROM visit");
+
+$test = new VisitTest();
+
+// create the visits and retreive
+$test->testCreate();
+$error = $test->testGetLabel();
+printf("createVisit() and getLabel() %s\n", $error ? "fail" : "pass");
+
+// create visit-project-subproject associations and retreive
+$test->testAddToProjectSubproject();
+$error = $test->testGetList();
+printf("getList() %s\n", ($error ? "fail" : "pass"));
+
+// modify visit label
+$error = $test->testSetLabel();
+printf("setLabel() %s\n", ($error ? "fail" : "pass"));
+
+
+// remove visit-project-subproject associations
+try {
+    $error = $test->testDeleteFromProject();
+}
+catch (LorisException $e) {
+    if ($e->getMessage() != "Need at least a project or subproject") {
+        print("** ERROR in deleting visit for a project or subproject\n");
+    }
+}
+
+printf("deleteFromProject() %s\n", ($error ? "fail" : "pass"));
+
+?>
```

Comment:
Safer to not include the closing tag in a PHP file, because then there's no chance of whitespace outside of the tag being accidentally significant.